File: examples/language_samples/sample.c
/* C sample file for syntax highlighting test */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SIZE 100
#define SQUARE(x) ((x) * (x))

typedef struct {
    int x;
    int y;
} Point;

typedef enum {
    STATUS_OK = 0,
    STATUS_ERROR = -1,
    STATUS_NOT_FOUND = -2
} Status;

/* Function prototypes */
Point create_point(int x, int y);
double distance(Point a, Point b);
int* create_array(size_t size);
void free_array(int* arr);

/* Create a new Point */
Point create_point(int x, int y) {
    Point p;
    p.x = x;
    p.y = y;
    return p;
}

/* Calculate Euclidean distance between two points */
double distance(Point a, Point b) {
    int dx = a.x - b.x;
    int dy = a.y - b.y;
    return sqrt((double)(dx * dx + dy * dy));
}

/* Dynamic array allocation */
int* create_array(size_t size) {
    int* arr = (int*)malloc(size * sizeof(int));
    if (arr == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        return NULL;
    }

    for (size_t i = 0; i < size; i++) {
        arr[i] = (int)i * 2;
    }

    return arr;
}

void free_array(int* arr) {
    if (arr != NULL) {
        free(arr);
    }
}

/* Linked list node */
struct Node {
    int data;
    struct Node* next;
};

struct Node* list_push(struct Node* head, int value) {
    struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
    if (new_node == NULL) {
        return head;
    }
    new_node->data = value;
    new_node->next = head;
    return new_node;
}

void list_print(struct Node* head) {
    struct Node* current = head;
    while (current != NULL) {
        printf("%d -> ", current->data);
        current = current->next;
    }
    printf("NULL\n");
}

int main(int argc, char* argv[]) {
    /* Point operations */
    Point p1 = create_point(0, 0);
    Point p2 = create_point(3, 4);

    printf("Point 1: (%d, %d)\n", p1.x, p1.y);
    printf("Point 2: (%d, %d)\n", p2.x, p2.y);
    printf("Distance: %.2f\n", distance(p1, p2));

    /* Array operations */
    int* numbers = create_array(10);
    if (numbers != NULL) {
        printf("Array: ");
        for (int i = 0; i < 10; i++) {
            printf("%d ", numbers[i]);
        }
        printf("\n");
        free_array(numbers);
    }

    /* Macro usage */
    int value = 5;
    printf("Square of %d is %d\n", value, SQUARE(value));

    /* Switch statement */
    Status status = STATUS_OK;
    switch (status) {
        case STATUS_OK:
            printf("Status: OK\n");
            break;
        case STATUS_ERROR:
            printf("Status: Error\n");
            break;
        default:
            printf("Status: Unknown\n");
            break;
    }

    return 0;
}

File: examples/language_samples/sample.zig
// Zig sample file for syntax highlighting test

const std = @import("std");

pub const Color = enum {
    red,
    green,
    blue,
};

pub const Point = struct {
    x: f32,
    y: f32,

    pub fn init(x: f32, y: f32) Point {
        return Point{ .x = x, .y = y };
    }

    pub fn distance(self: Point, other: Point) f32 {
        const dx = self.x - other.x;
        const dy = self.y - other.y;
        return @sqrt(dx * dx + dy * dy);
    }
};

const LinkedList = struct {
    const Node = struct {
        data: i32,
        next: ?*Node,
    };

    head: ?*Node,
    allocator: std.mem.Allocator,

    pub fn init(allocator: std.mem.Allocator) LinkedList {
        return LinkedList{
            .head = null,
            .allocator = allocator,
        };
    }

    pub fn push(self: *LinkedList, value: i32) !void {
        const node = try self.allocator.create(Node);
        node.* = Node{
            .data = value,
            .next = self.head,
        };
        self.head = node;
    }
};

fn fibonacci(n: u32) u64 {
    if (n <= 1) return n;

    var prev: u64 = 0;
    var curr: u64 = 1;
    var i: u32 = 2;

    while (i <= n) : (i += 1) {
        const next = prev + curr;
        prev = curr;
        curr = next;
    }

    return curr;
}

pub fn main() !void {
    const stdout = std.io.getStdOut().writer();

    // Print fibonacci numbers
    var i: u32 = 0;
    while (i < 10) : (i += 1) {
        try stdout.print("fib({d}) = {d}\n", .{ i, fibonacci(i) });
    }

    // Use Point struct
    const p1 = Point.init(0.0, 0.0);
    const p2 = Point.init(3.0, 4.0);
    try stdout.print("Distance: {d}\n", .{p1.distance(p2)});

    // Color enum
    const color: Color = .green;
    switch (color) {
        .red => try stdout.print("Red!\n", .{}),
        .green => try stdout.print("Green!\n", .{}),
        .blue => try stdout.print("Blue!\n", .{}),
    }
}

test "fibonacci test" {
    try std.testing.expectEqual(@as(u64, 0), fibonacci(0));
    try std.testing.expectEqual(@as(u64, 1), fibonacci(1));
    try std.testing.expectEqual(@as(u64, 55), fibonacci(10));
}
